### PR TITLE
DDF-2869 Rework ConfigurationFilesPoller to be more reliable

### PR DESCRIPTION
**Pull request by brendan-hofmann**
#### What does this PR do?
Replace nio with commons-io's file watcher, and replace the one second sleep with a poll for the file size to become stable.

These changes should mitigate itest stability problems in CI environments that are prone to IOWait spikes.

Also includes some improvements to itest logging that were needed to debug this solution anyways.
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@ahoffer @clockard 
#### Select at least one member from relevant component team(s) from below (at least one component team member needs to approve the PR).
[Continuous Integration](https://github.com/orgs/codice/teams/continuous-integration)
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@coyotesqrl
@lessarderic
#### What are the relevant tickets?
[DDF-2869](https://codice.atlassian.net/browse/DDF-2869)
#### Checklist:
- [ ] Documentation Updated
- [ ] Change Log Updated
- [x] Update / Add Unit Tests
- [x] Update / Add Integration Tests

### CODE REVIEW

File: distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/ServiceManagerImpl.java [hunk 1]
```diff
@@ -681,27 +680,36 @@ public void printInactiveBundlesInfo() {
...
-                Enumeration<String> keys = headers.keys();
-
-                while (keys.hasMoreElements()) {
-                    String key = keys.nextElement();
-                    headerString.append(key)
-                            .append("=")
-                            .append(headers.get(key))
-                            .append(", ");
-                }
+                if (bundleService.getInfo(bundle)
```
Comment (brendan-hofmann) [inline distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/ServiceManagerImpl.java]: This section was donated by @coyotesqrl to improve bundle fragment logging.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java [hunk 2]
```diff
@@ -1907,6 +1918,7 @@ public void testContentDirectoryMonitor() throws Exception {
         cdmProperties.putAll(getServiceManager().getMetatypeDefaults("content-core-directorymonitor",
                 "org.codice.ddf.catalog.content.monitor.ContentDirectoryMonitor"));
         cdmProperties.put("monitoredDirectoryPath", tmpDir.toString() + "/");
+        cdmProperties.put("processingMechanism", "delete");
```
Comment (brendan-hofmann) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java]: Interestingly enough, I found this one by accident. When the default processing mechanism was changed we didn't update the itests, so as a result we had no itest coverage for this mechanism.

**Comment by clockard** [comment]
Looks like there are some unused imports that is causing the build to fail on checkstyle.

File: distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/annotations/PaxExamRule.java [hunk 3]
```diff
@@ -90,7 +98,9 @@ private void starting(Description description) {
             runAnnotations(PostTestConstruct.class, testClass);
         } catch (Throwable throwable) {
             setupFailed = true;
-            fail(String.format(BEFORE_EXAM_FAILURE_MESSAGE, testClassName, throwable.getMessage()));
+            StringWriter errors = new StringWriter();
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/annotations/PaxExamRule.java]: Could refactor this block as it's duped three times.

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/PaxExamRuleTest.java [hunk 4]
```diff
@@ -175,9 +176,13 @@ public void failingBeforeExam() {
         JUnitCore core = new JUnitCore();
         Result result = core.run(FailingBeforeExamTest.class);
 
-        assertThat(result.getFailures()).extracting("message")
-                .containsOnly(EXPECTED_BEFORE_EXAM_ERROR_MESSAGE,
-                        PaxExamRule.EXAM_SETUP_FAILED_MESSAGE);
+        assertThat(result.getFailures()
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/PaxExamRuleTest.java]: This formatting is terrible. Maybe extract the two boolean terms into variables so it becomes more readable?

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/PaxExamRuleTest.java [hunk 5]
```diff
@@ -187,9 +192,13 @@ public void failingAfterExam() {
         JUnitCore core = new JUnitCore();
         Result result = core.run(FailingAfterExamTest.class);
 
-        assertThat(result.getFailures()).extracting("message")
-                .contains(EXPECTED_AFTER_EXAM_ERROR_MESSAGE, FAILING_TEST_MESSAGE);
-
+        assertThat(result.getFailures()
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/PaxExamRuleTest.java]: Same formatting content. It would be fine if you could move the second boolean term to start on the line after the first so it lined up better but in the absence of that, I think two terms will be easier to read.

File: distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/annotations/PaxExamRule.java (hunk 3 above)
Comment (brendan-hofmann) [inline distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/annotations/PaxExamRule.java]: Got it, guess that didn't get refactored because of the `String.format()`.

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/PaxExamRuleTest.java (hunk 4 above)
Comment (brendan-hofmann) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/PaxExamRuleTest.java]: Can do.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java [hunk 6]
```diff
@@ -198,8 +198,8 @@ public void tearDown() throws IOException {
     @Test
     public void testCreateStorage() throws IOException {
         String fileName = testName.getMethodName() + ".jpg";
-        File tmpFile = createTemporaryFile(fileName, IOUtils.toInputStream(getFileContent(
-                SAMPLE_IMAGE)));
+        File tmpFile = createTemporaryFile(fileName,
+                IOUtils.toInputStream(getFileContent(SAMPLE_IMAGE)));
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java]: ❓ There is a lot of "reformatting noise" in this diff. Could we just have the actual changes in this diff and keep the reformatting changes in a separate commit/PR?

Comment (brendan-hofmann) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java]: They're pretty entangled now. You can tell github to ignore whitespace in the diff and that helps.

File: distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/annotations/PaxExamRule.java [hunk 7]
```diff
@@ -62,6 +64,12 @@ public PaxExamRule(Object testInstance) {
         this.testInstance = testInstance;
     }
 
+    public static void failWithThrowableStacktrace(Throwable t, String message) {
```
Comment (brendan-hofmann) [inline distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/annotations/PaxExamRule.java]: I botched it a bit in my last refactoring, but the intent was to avoid coupling in the other direction, since `PaxExamRule` also should use this helper method, so I was trying to avoid a circular dependency. I can create a separate class just for this method if you'd prefer.

File: distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/ServiceManagerImpl.java [hunk 8]
```diff
@@ -681,27 +680,36 @@ public void printInactiveBundlesInfo() {
 
     private void printInactiveBundles(Consumer<String> headerConsumer,
             BiConsumer<String, Object[]> logConsumer) {
+        if (bundleService == null) {
```
Comment (brendan-hofmann) [inline distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/ServiceManagerImpl.java]: Those print statements are called fairly often during itest setup. I think the member variable makes sense to avoid having to retrieve the reference for every call. I don't think this class is touched very often, but since there are only two instances I am comfortable with leaving it explicit for now unless you feel strongly otherwise.

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java [hunk 9]
```diff
@@ -80,16 +72,7 @@ public void init() throws IOException {
...
-            LOGGER.debug("Registering path [{}] with Watch Service.",
-                    configurationDirectoryPath.toString());
-            configurationDirectoryPath.register(watchService, ENTRY_CREATE);
-        } catch (IOException e) {
-            LOGGER.error("Unable to register path [{}] with Watch Service",
-                    configurationDirectoryPath.toString(),
-                    e);
-            throw e;
-        }
+        watchService.addListener(new ListenerImpl());
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: ❗️ Does this get the polling started or does the polling start the first time `checkAndNotify()` is called on L90? I'm asking because we need to make sure we don't process the same configuration file twice, once because `processExistingConfigurationFiles()` is called in the `register()` method and once because `ListenerImpl` gets notified.

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java [hunk 10]
```diff
@@ -102,67 +85,15 @@ public void register(@NotNull ChangeListener listener) {
...
-        } catch (InterruptedException | RuntimeException e) {
-            LOGGER.debug("The [{}] was interrupted.",
-                    this.getClass()
-                            .getName(),
-                    e);
-            Thread.currentThread()
-                    .interrupt();
+        while (!Thread.currentThread()
+                .isInterrupted()) {
+            watchService.checkAndNotify(); // blocking
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: ❓ Is this really a blocking call? It's not clear from the documentation. If not, we may want to have some kind of sleep between `checkAndNotify` calls to prevent a busy loop.

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java (hunk 9 above)
Comment (lessarderic) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: Actually, I wonder if we shouldn't just move the `addListener` call and `executorService.execute()` call inside the `register()` method and use the listener's `onStart` method to process existing files. This would guarantee that the processing of existing files and the notifications are all done on the same thread and should eliminate the risk of having a file being processed twice.

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java [hunk 11]
```diff
@@ -58,19 +53,16 @@
...
-            @NotNull ExecutorService executorService) {
+            @NotNull String fileExtension, @NotNull ExecutorService executorService) {
         notNull(configurationDirectoryPath, "configurationDirectoryPath cannot be null");
         notNull(fileExtension, "fileExtension cannot be null");
-        notNull(watchService, "watchService cannot be null");
         notNull(executorService, "executorService cannot be null");
 
         this.configurationDirectoryPath = configurationDirectoryPath;
-        this.watchService = watchService;
+        this.watchService = new FileAlterationObserver(configurationDirectoryPath.toString());
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: ❓ Do you think we could have used the `FileAlterationMonitor` to replace this whole class?

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java [hunk 12]
```diff
@@ -199,4 +130,82 @@ private void processExistingConfigurationFiles() {
...
+        public void onDirectoryDelete(File file) {
+        }
+
+        @Override
+        public void onFileCreate(File file) {
+            try {
+                String filename = file.getName();
+                LOGGER.debug("Watcher has been notified. Handling event for [{}].", filename);
+
+                if (!filename.endsWith(fileExtension)) {
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: ❓ Couldn't we use a `IOFileFilter` when we create the `FileAlterationObserver` instead of doing this every time this method gets called?

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java [hunk 13]
```diff
@@ -199,4 +130,82 @@ private void processExistingConfigurationFiles() {
...
+                if (changeListener != null) {
+                    waitForFileToBeCompletelyWritten(file);
+                    LOGGER.debug("Notifying [{}] of creation for file [{}].",
+                            changeListener.getClass()
+                                    .getName(),
+                            configurationDirectoryPath.resolve(filename));
+                    changeListener.notify(configurationDirectoryPath.resolve(filename));
+                }
+            } catch (InterruptedException e) {
+                LOGGER.debug("The [{}] was interrupted.",
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: ❗️ This is pretty bad if it happens because configuration files will not be processed anymore. We may want to bump up the logging level here (WARN?) and say that the application/bundle needs be restarted.

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java [hunk 14]
```diff
@@ -102,67 +85,15 @@ public void register(@NotNull ChangeListener listener) {
...
-            LOGGER.debug("The [{}] was interrupted.",
-                    this.getClass()
-                            .getName(),
-                    e);
-            Thread.currentThread()
-                    .interrupt();
+        while (!Thread.currentThread()
+                .isInterrupted()) {
+            watchService.checkAndNotify(); // blocking
         }
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: ❓ Wonder if we should log something when we exit this loop to let admins know that file polling has stopped and that the application/bundle needs to be restarted if the loop stopped because the thread was interrupted but the application is still running.

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java [hunk 15]
```diff
@@ -199,4 +130,82 @@ private void processExistingConfigurationFiles() {
...
+                        .interrupt();
+            }
+        }
+
+        private void waitForFileToBeCompletelyWritten(File file) throws InterruptedException {
+            Long fileSizeBefore = 0L;
+            Long fileSizeAfter = 0L;
+            do {
+                fileSizeBefore = file.length();
+                Thread.sleep(2000);
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: ❗️ We should use a constant instead of hardcoding `2000` here and on L192.

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java [hunk 16]
```diff
@@ -199,4 +130,82 @@ private void processExistingConfigurationFiles() {
...
+                        this.getClass()
+                                .getName(),
+                        e);
+                Thread.currentThread()
+                        .interrupt();
+            }
+        }
+
+        private void waitForFileToBeCompletelyWritten(File file) throws InterruptedException {
+            Long fileSizeBefore = 0L;
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: ❓ `long` instead of `Long`?

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java (hunk 11 above)
Comment (brendan-hofmann) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: I'll see if we can, I think it should be possible.

File: distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/annotations/PaxExamRule.java (hunk 7 above)
Comment (lessarderic) [inline distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/annotations/PaxExamRule.java]: Good point, didn't think of that. A new common logging class seems like a good way to go. Pretty sure we'll add new logging related methods in there before long :-)

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/platform/TestConfiguration.java [hunk 17]
```diff
@@ -680,16 +679,17 @@ public void testConfigStatusFailedImports() throws Exception {
...
-                getFileContentAsStream(INVALID_CONFIG_FILE_2));
         String output = console.runCommand(STATUS_COMMAND);
-
         assertThat(output,
                 containsString(String.format(FAILED_IMPORT_MESSAGE, INVALID_CONFIG_FILE_1)));
-        assertThat(output,
-                containsString(String.format(FAILED_IMPORT_MESSAGE, INVALID_CONFIG_FILE_2)));
         assertThat(Files.exists(getPathToFailedDirectory().resolve(INVALID_CONFIG_FILE_1)),
                 is(true));
+
```
Comment (brendan-hofmann) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/platform/TestConfiguration.java]: I did not know the ordering was significant, so I reordered them to make the test pass. I will revert this change and investigate further.

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java (hunk 12 above)
Comment (brendan-hofmann) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: `SuffixFileFilter` is even more convenient. 😉

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java [hunk 18]
```diff
@@ -14,38 +14,34 @@
...
 /**
  * Class that monitors changes to files with a specific extension in a directory.
  */
-public class ConfigurationFilesPoller implements Runnable {
+public class ConfigurationFilesPoller implements FileAlterationListener {
+
     private static final Logger LOGGER = LoggerFactory.getLogger(ConfigurationFilesPoller.class);
 
-    private final WatchService watchService;
+    public static final long INTERVAL = 2000L;
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: ✏️ `POLLING_INTERVAL`?

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java [hunk 19]
```diff
@@ -58,100 +54,65 @@
...
-            configurationDirectoryPath.register(watchService, ENTRY_CREATE);
-        } catch (IOException e) {
-            LOGGER.error("Unable to register path [{}] with Watch Service",
-                    configurationDirectoryPath.toString(),
-                    e);
-            throw e;
+            watchService.start();
+        } catch (Exception e) {
+            String msg = "Failed to start, 'Platform :: Migration' must be restarted: ";
+            LOGGER.info(msg, e);
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: ❓If we really want the stack trace at a different log level, should we use `DEBUG` instead of `INFO`?

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java [hunk 20]
```diff
@@ -58,100 +54,65 @@
...
-                // (possibly deleted)
-                if (!key.reset()) {
-                    LOGGER.warn("Configurations in [{}] are no longer able to be observed.",
-                            configurationDirectoryPath.toString());
-                    break;
-                }
+            String filename = file.getName();
+            LOGGER.debug("Watcher has been notified. Handling event for [{}].", filename);
+
+            if (changeListener != null) {
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: ❓Since these methods will only be called when `register` is called and `changeListener` is set, do we still need this check?

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java [hunk 21]
```diff
@@ -58,100 +54,65 @@
...
+                waitForFileToBeCompletelyWritten(file);
+                LOGGER.debug("Notifying [{}] of creation for file [{}].",
+                        changeListener.getClass()
+                                .getName(),
+                        file.getCanonicalPath());
+                changeListener.notify(Paths.get(file.getCanonicalPath()));
             }
-        } catch (InterruptedException | RuntimeException e) {
-            LOGGER.debug("The [{}] was interrupted.",
+        } catch (InterruptedException | IOException e) {
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: ❓Wonder if we should split this catch into `InterruptedException` and `IOException`. In the former we could do what we do now, but in the latter we should be able to log something and keep going. We should also consider what to do with `RuntimeException`. Thoughts?

File: platform/migration/platform-migration/src/test/java/org/codice/ddf/configuration/admin/ConfigurationFilesPollerTest.java [hunk 22]
```diff
@@ -86,234 +77,40 @@ public void setup() {
...
-        configurationFilesPoller.register(mockChangeListener);
+        Path mockPath = Files.createTempDirectory("foo");
+        ConfigurationFilesPoller configurationFilesPoller = new ConfigurationFilesPoller(mockPath,
+                FILE_EXT);
 
         // Perform Test
-        configurationFilesPoller.run();
+        configurationFilesPoller.register(mockChangeListener);
+        Files.createTempFile(mockPath, "foo", FILE_EXT);
+        Thread.sleep(5000);
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/test/java/org/codice/ddf/configuration/admin/ConfigurationFilesPollerTest.java]: ❗️Instead of testing with real files and having to deal with sleeps and such, couldn't we simply test the individual notification methods and make sure they do what they're supposed to? That may require injecting and mocking out the `FileAlterationMonitor` object, but that sounds like a better option imho.

File: distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/utils/LoggingUtils.java [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+import static org.junit.Assert.fail;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+
+public class LoggingUtils {
+    public static void failWithThrowableStacktrace(Throwable t, String message) {
+        StringWriter errors = new StringWriter();
+        t.printStackTrace(new PrintWriter(errors));
+        fail(message + errors.toString());
```
Comment (brendan-hofmann) [inline distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/utils/LoggingUtils.java]: I don't think there's a one-size-fits-all answer, so I'd rather leave that up to the caller.

Comment (lessarderic) [inline distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/utils/LoggingUtils.java]: I'm ok with leaving that up to the caller as long as we (JavaDoc) tell them they need to do it. Otherwise we'll end up with a bunch of log messages that don't have a space between `message` and `errors`.

File: platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java [hunk 24]
```diff
@@ -58,145 +53,96 @@
...
-        this.executorService = executorService;
         this.fileExtension = fileExtension;
     }
 
-    public void init() throws IOException {
-        LOGGER.debug("Starting {}...",
-                this.getClass()
-                        .getName());
-
+    public void register(@NotNull ChangeListener listener) {
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ConfigurationFilesPoller.java]: ✏️  Just noticed that this class uses the `@NotNull` annotation. We changed our convention since we wrote this class to assume nothing should be `null` unless we use `@Nullable`. Feel free to remove those if you want.

File: platform/migration/platform-migration/src/test/java/org/codice/ddf/configuration/admin/ConfigurationFilesPollerTest.java [hunk 25]
```diff
@@ -86,362 +76,198 @@ public void setup() {
...
 
         // Perform Test
         configurationFilesPoller.register(mockChangeListener);
+        configurationFilesPoller.onFileCreate(file);
 
         // Verify
-        verify(mockChangeListener).notify(configFilePath);
-        verify(mockChangeListener, never()).notify(directoryPath);
-        verify(mockChangeListener, never()).notify(nonConfigFilePath);
+        verify(mockChangeListener).notify(any());
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/test/java/org/codice/ddf/configuration/admin/ConfigurationFilesPollerTest.java]: ❗️ We should avoid using `any()` here and use a specific matcher instead, otherwise we're not validating that the listener is called with the proper file.

File: platform/migration/platform-migration/src/test/java/org/codice/ddf/configuration/admin/ConfigurationFilesPollerTest.java [hunk 26]
```diff
@@ -86,362 +76,198 @@ public void setup() {
...
-        List<WatchEvent<?>> watchEvents = getSingleMockWatchEvent(mockWatchEvent);
-        WatchKey mockWatchKey = getMockWatchKey(watchEvents);
-        WatchService mockWatchService = getMockWatchService(mockWatchKey);
-        Path mockConfigurationDirectory = getMockConfigurationDirectoryPath(mockPath,
-                mock(Path.class));
-        ConfigurationFilesPoller configurationFilesPoller = createConfigurationFilesPoller(
-                mockWatchService,
-                mockConfigurationDirectory);
-        configurationFilesPoller.register(mockChangeListener);
+        ConfigurationFilesPoller configurationFilesPoller = new ConfigurationFilesPoller(
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/test/java/org/codice/ddf/configuration/admin/ConfigurationFilesPollerTest.java]: ✏️  Instead of inlining this class everywhere, could we create a static inner class and use that instead?

File: platform/migration/platform-migration/src/test/java/org/codice/ddf/configuration/admin/ConfigurationFilesPollerTest.java [hunk 27]
```diff
@@ -86,362 +76,198 @@ public void setup() {
...
+                configFilePath,
+                FILE_EXT) {
+            @Override
+            void waitForFileToBeCompletelyWritten(File file) throws InterruptedException {
+                return;
+            }
+        };
+        when(configFilePath.toAbsolutePath()).thenReturn(Paths.get("foo"));
+        File file = Files.createTempFile("foo", "bar")
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/test/java/org/codice/ddf/configuration/admin/ConfigurationFilesPollerTest.java]: ✏️  Could we move this in a `@Before` method?

File: platform/migration/platform-migration/src/test/java/org/codice/ddf/configuration/admin/ConfigurationFilesPollerTest.java [hunk 28]
```diff
@@ -86,362 +76,198 @@ public void setup() {
...
-                mock(Path.class));
         ConfigurationFilesPoller configurationFilesPoller = new ConfigurationFilesPoller(
-                mockConfigurationDirectory,
-                FILE_EXT,
-                mockWatchService,
-                mockExecutorService);
+                configFilePath,
+                FILE_EXT);
+        when(configFilePath.toAbsolutePath()).thenReturn(Paths.get("foo"));
+        // will cause an NPE in Paths.get()
```
Comment (lessarderic) [inline platform/migration/platform-migration/src/test/java/org/codice/ddf/configuration/admin/ConfigurationFilesPollerTest.java]: ❓ Could we do `when(file.getPath()).thenThrow(new RuntimeException())` instead of relying on `Paths.get()` throwing an exception? Another option would be to overwrite `waitForFileToBeCompletelyWritten()` and have it throw the `RuntimeException`.

**Comment by brendan-hofmann** [comment]
@lessarderic I liked the template approach better. I refactored the implementation and added unit tests accordingly.
